### DISCUSSION TITLE
How to use with other contents like just a partial winbar

### DISCUSSION THREAD

**Discussion by weilbith** [Bekaboo/dropbar.nvim#9] [General]
Hey 👋🏾 
I like your plugin. So far I implemented this manually (though with some support) using Heirline. But my winbar does not include the "breakcumbs" only, but more content. So my intention is to use your winbar output, but manually put it into my winbar (or rather with Heirline). 

I've seen your great API documentation. And I thought the following sequence would do it:

```lua
require('dropbar.configs').set()
require('dropbar.hlgroups').init()

-- Use require('dropbar.api').get_current_dropbar() now in Heirline.
```

But it doesn't. The first two calls already register the winbar and overwrite my custom one. My other segments for the tests etc. are no more visible.

Could you please help me to make this work?

EDIT:
I'm quite confused. Even if I provide this option table to the `configs.set` function, it still registers itself.

```lua
{ enable = false }
```

**Comment by Bekaboo** [comment]
> `{ enable = false }`

Maybe try `general = { enable = false }`? :)

**Reply by weilbith** [reply]
Good point. 😆 
But now my custom winbar pop-ups and then gets overwritten with an empty one... 🙈 
I guess I first have to read your code before I can use it as a single segment in my windbar. 😕

**Reply by Bekaboo** [reply]
> I guess I first have to read your code before I can use it as a single segment in my windbar.

You haven't to although you are welcomed to.

> But now my custom winbar pop-ups and then gets overwritten with an empty one..

I think I already have a answer in mind about what is happening here, this is an unexpected behavior, will work to fix it later.

**Reply by weilbith** [reply]
I saw in the code that if `enable` is false, it will set `vim.wo.winbar` to `nil`. That is causing the issue for me. If I comment that out locally, I can call the `setup` and then in my own winbar I call `_G.dropbar.get_dropbar_str()`.
Now I'm only left with an LSP data access issue by dropbar I do not understand yet. Because it doesn't appear if I use it out of the box.
`dropbar.nvim/lua/dropbar/sources/lsp.lua:408: attempt to index field 'data'  (a nil value)`

**Reply by Bekaboo** [reply]
You are right, glad that you find it yourself. `data` field is only for extra misc data, maybe the `symbol` field is what you are looking for.

**Reply by weilbith** [reply]
I'm actually looking for nothing. 😆 
Just trying to figure out why it is failing. I suspect that is the response handler callback. So that `data` should be from the server response, correct? Not sure why this happens. It is also independent of the server/language. And after the first initial error, it just works like charm. 🤷🏾

**Reply by Bekaboo** [reply]
Actually I have no clear idea about what you want to achieve in your config and what you want from this plugin, but if you are trying to get lsp data you are going to look into the `symbol` field instead of the `data` field.

**Reply by weilbith** [reply]
All good. Realized it is the `LspAttach` auto-command. Stupid.

What I want to achieve:
Your plugin is written in a way that is manages the full winbar. I intend to use your winbar just a part of my winbar. I use a plugin called Heirline which just gives you a little nice framework to write your own winbar composed of what every you like. And the breadcumbs stuff you have written here is nice. So I would like to replace mine with yours. But my winbar consists of more than these breadcrumbs. For example do I also see the number succeed/failed/total test cases if this is a test file. 
So what I try to achieve is accessing your plugins in a way that it is not fully taking over my winbar, but I can embed it into my custom one as one part of it.

**Reply by weilbith** [reply]
And my technical issue is just a case that I can see in the logs an `LspAttach` autocommand trigger that has no `data` property and thereby no `client_id`. It would be easy to make this more robust and open a PR to ignore such events. But I would first like to figure out why there is such an event on my setup./

So I always gets "number of server + 1" events of kind `LspAttach`. The first one looks like that:
```lua
{
  buf = 1,
  event = "LspAttach",
  file = "*",
  group = 75,
  id = 126,
  match = "/home/user_name/projects/project_name/*"
}
```

Not sure if that is normal. All other events of that kind match to a server each. So ignoring it would be just fine and makes the code more "robust". Though I'm still confused.

**Reply by Bekaboo** [reply]
That's much clearer. If you don't mind I think it will be much more easier to make your custom winbar a source and feed it to dropbar, see [here](https://github.com/Bekaboo/dropbar.nvim#dropbar_source_t) and [here](https://github.com/Bekaboo/dropbar.nvim#making-a-new-source). If you are interested I hope you can submit a PR to include your souce as part of dropbar's built-in sources ;)

**Reply by Bekaboo** [reply]
I believe the in latest commit 5d7030f88813a53200b23c71b906f9bbb8934019 `dropbar.nvim` will not override the winbar when `opts.general.enable` is evaluated to false, please check it out and let me know if it works for you

**Reply by JuanZoran** [reply]
Actually, there are lots of requirements in winbar
ex: I am using lualine winbar to display the current time, memory usage, and attached lsp name, I wonder to know that how can I migrate that to dropbar.nvim
![image](https://github.com/Bekaboo/dropbar.nvim/assets/107862700/3ff46b1d-d705-48a8-82bb-e473b6a4a105)

**Reply by weilbith** [reply]
> I believe the in latest commit https://github.com/Bekaboo/dropbar.nvim/commit/5d7030f88813a53200b23c71b906f9bbb8934019 dropbar.nvim will not override the winbar when opts.general.enable is evaluated to false, please check it out and let me know if it works for you

Thank you very much! That does the job for me. I could remove my local changes. Thank you very much!

**Reply by weilbith** [reply]
Sorry, I know this discussion got out of control. But I just wanna share a "feature" I implemented myself, which might be interesting to get support by you if you think it is interesting.
So I actually liked the idea to have more colors in the winbar as I had before. So I decided to use the highlighting not only on the icon, but also the name/text. **But** except the arrow/caret. So I basically just did a string regular expression replacement to achieve this. 😆 
The code looks as ugly as that (note that the variable `separator` is also used in the setup configuration):
```lua
local content = _G.dropbar.get_dropbar_str()
content = content:gsub('%%%*', '')
content = content:gsub(separator, '%%#WhiteBold#' .. separator)
```

Which then looks like that:
Sorry, I know this discussion got out of control. But I just wanna share a "feature" I implemented myself, which might be interesting to get support by you if you think it is interesting.
So I actually liked the idea to have more colors in the winbar as I had before. So I decided to use the highlighting not only on the icon, but also the name/text. **But** except the arrow/caret. So I basically just did a string regular expression replacement to achieve this. 😆 
The code looks as ugly as that (note that the variable `separator` is also used in the setup configuration):
```lua
local content = _G.dropbar.get_dropbar_str()
content = content:gsub('%%%*', '')
content = content:gsub(separator, '%%#WhiteBold#' .. separator)
```

Which then looks like that:
![image](https://github.com/Bekaboo/dropbar.nvim/assets/12543647/68af618a-285c-4ec1-b19e-ca43359555c5)

**Reply by Bekaboo** [reply]
@weilbith Text colors support is planned in #18

**Reply by weilbith** [reply]
Cool. I'm subscribed to the ticket now. :+1:
Thanks for all your work!

**Comment by Bekaboo** [comment]
@JuanZoran 

The main development of `dropbar.nvim` revolves around information with a tree-like structure (lsp/treesitter symbols, markdown heading hierarchies, and directory hierarchies, etc) and a drop-down menu to visualize the tree structure, making the tree more accessible.

From your screenshot I can see you use the winbar to show information about the date, time, and memory usage, etc -- that is very different kinds of information compared with what `dropbar.nvim` aims for, it is possible to show these information in `dropar.nvim` but I doubt if you can get any real benefits from switching to `dropbar.nvim`.

**Reply by JuanZoran** [reply]
I think the other solution is that dropbar.nvim can export the winbar module and handlers, so that dropbar.nvim can be anywhere

**Reply by JuanZoran** [reply]
That's what navic.nvim is doing... I am using navic.nvim but I think dropbar is more cool

**Reply by Bekaboo** [reply]
> I think the other solution is that dropbar.nvim can export the winbar module and handlers, so that dropbar.nvim can be anywhere

If I'm not misunderstanding you, are you going to embed `dropbar.nvim` as a source of your lualine winbar?

**Reply by JuanZoran** [reply]
Yes, I would like to do that

**Reply by Bekaboo** [reply]
Then I think you might want to set `opts.general.enable = false` and use `_G.dropbar.get_dropbar_str()` to get the string representation of the winbar with drop-down menu from dropbar.nvim

**Reply by JuanZoran** [reply]
I don't know if I don't setup the dropbar.nvim correctly, but the `require'dropbar.api'.pick()` doesn't work sometimes....


if you would like to check my config for that, here is my lualine config:

```lua
local function memory_use()
    local use = (1 - (vim.loop.get_free_memory() / vim.loop.get_total_memory())) * 100
    return (' Memory: %.2f'):format(use) .. ' %%'
end

local function opts()
    local winbar = {
        lualine_a = {
            function() return '盛 ' .. os.date '%A %H:%M' end,
        },
        lualine_b = {
            { 'fancy_diagnostics' },
            {
                require 'noice'.api.status.mode.get,
                cond = require 'noice'.api.status.mode.has,
                color = { fg = '#ff9e64' },
            },
        },
        lualine_c = {
            {
                -- require 'nvim-navic'.get_location,
                -- cond = require 'nvim-navic'.is_available,
                -- function() return require 'nvim-navic'.get_location() end, -- FIXME : why it is not working
                -- cond = require 'nvim-navic'.is_available,
                function()
                    return dropbar.get_dropbar_str()
                end,
            },
        },
        lualine_y = {
            {
                memory_use,
                color = {
                    fg = '#69bbae',
                    -- bg = '#1e1e2e'
                    bg = '#304263',
                },
            },
        },
        lualine_z = {
            {
                'fancy_lsp_servers',
                color = {
                    fg = '#1e1e2e',
                    bg = '#986FEC',
                },
            },
        },
    }

    local sections = {
        lualine_a = {
            { 'fancy_mode', width = 6 },
        },
        lualine_b = {
            {
                'b:gitsigns_head',
                icon = ' ',
                color = {
                    fg = '#69bbae',
                    bg = '#1e1e2e',
                },
            },
            { 'fancy_diff' },
        },
        lualine_c = {
            { 'fancy_cwd', substitute_home = true },
        },
    }

    return {
        options = {
            component_separators = { left = '', right = '' },
            theme = 'auto',
            section_separators = {
                right = '',
                left = '',
            },
            globalstatus = true,
            disabled_filetypes = {
                statusline = {
                    'alpha',
                    'toggleterm',
                    'lspsagaoutline',
                },
                winbar = {
                    'alpha',
                    'help',
                    'lspsagaoutline',
                    'Trouble',
                    'toggleterm',
                    'dap-repl',
                    'dapui_console',
                    'dapui_watches',
                    'dapui_stacks',
                    'dapui_breakpoints',
                    'dapui_scopes',
                },
            },
        },
        sections = sections,
        winbar = winbar,
        extensions = { 'trouble', 'lazy', 'neo-tree' },
    }
end


return {
    'nvim-lualine/lualine.nvim',
    event = 'VeryLazy',
    dependencies = {
        'meuter/lualine-so-fancy.nvim',
        {
            'Bekaboo/dropbar.nvim',
            keys = {
                { '<leader>k', function() require 'dropbar.api'.pick() end },
            },
            opts = {
                general = { enable = false },
            },
        },
    },
    opts = opts,
}

```

**Reply by Bekaboo** [reply]
> require'dropbar.api'.pick() doesn't work sometimes....

Please open a separate issue for that and follow the instructions in the bug report template.

**Reply by JuanZoran** [reply]
https://github.com/Bekaboo/dropbar.nvim/assets/107862700/6e1eecc7-4afa-44cf-af9a-145528d0c588

you can see that when I press that key which I map to `require'dropbar.api'.pick()`, it doesn't show those characters but if I press any key other it will show up just a moment
